Add support for multiple comparison columns (i.e. one comparison column per producer sinking to a table)

egalpin opened the pull request:
The idea of this change set is to allow for defining multiple comparison columns to better support use cases where multiple disparate producer groups are sinking data to the same table.  Ex. In the case of partial upsert, if 3 producer groups (applications) produce data for 3 mutually exclusive sets of columns (aside from primary key), maintaining a comparison value for each producer group allows for safely updating different "parts" of a complete row without the need to synchronize timestamps across producer groups.

The intention of this change set is definitely to maintain full backward compatibility, so if it does not that should be called out.

`tag`: feature
`release notes`:  for `upsertConfig`, `comparisonColumn` or `comparisonColumns` can be used (only the latter in the case where an array of columns is being set.

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:
```diff
@@ -51,8 +53,8 @@ public enum Strategy {
-  @JsonPropertyDescription("Column for upsert comparison, default to time column")
-  private String _comparisonColumn;
+  @JsonPropertyDescription("Columns for upsert comparison, default to time column")
+  private List<String> _comparisonColumns;
```
This is backward incompatible change. We should support both `comparisonColumn` (existing single comparison column) and `comparisonColumns` (new added multiple comparison columns), and only allow one of them to be configured (at least one of them should be `null`, both of them can be `null` and we use time column as the comparison column)

egalpin replied:
As a result of having a setter for both a String or list of String (below code snippet), Jackson databind will use the correct setter.  In this case, a singular `comparisonColumn` will be coerced to a list, so all existing configs will continue to work as they did before without the need for any users to alter their configs.  I have a test for that using [pinot-common/src/test/resources/testConfigs/multipleIndexPartialUpsertConfig.json](https://github.com/apache/pinot/pull/10234/files#diff-d7f3fdcb8570dfc02a7430ebf0ac83051edd37a1bc5c80a2e728fec0122b3206) 

```java
  public void setComparisonColumn(List<String> comparisonColumns) {
    _comparisonColumns = comparisonColumns;
  }

  public void setComparisonColumn(String comparisonColumn) {
    _comparisonColumns = Collections.singletonList(comparisonColumn);
  }
  ```
  
  I think there might be another question around whether or not we want to do that coercion from String to SingletonList.  If we don't coerce, I feel there will be a lot of conditional logic throughout the codebase.  If we do coerce, there will be some overhead for cases with only one comparisonColumn.
  
There shouldn't really be any performance impact since list of size 1 will have O(1) complexity, but there would be memory overhead in the use of ComparisonColumns class. One thing that's nice about coercing to always have a list of columns is that it would mean anyone could add a new comparisonColumn at any time without the need for a server restart

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/RecordInfo.java:
```diff
@@ -18,19 +18,27 @@
   private final PrimaryKey _primaryKey;
   private final int _docId;
   private final Comparable _comparisonValue;
+  private final String _comparisonColumnName;
```
Adding this field will add extra memory footprint to the existing use cases. I think this info can be derived from the `_comparisonValue`, and we should add a special comparator for the new `_comparisonValue`

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/UpsertUtils.java:
```diff
@@ -134,6 +137,38 @@ public void close()
     }
   }
 
+  public static class ComparisonColumnReader implements Closeable {
```
Let's make 2 reader classes, one for single column and one for multiple columns to avoid impacting the performance for existing use cases

Jackie-Jiang reviewed (commented):
The approach looks good in general. We want to avoid adding overhead for single comparison column case, so let's try to avoid using list (at least avoid per-entry list) when there is only one comparison column

egalpin commented:
@Jackie-Jiang If we handle the case where there is a single comparisonColumn in a special way such that the column name is not stored on the RecordLocation, would we be able to support the case where a user wants to update to having many columns after previously having a single comparisonColumn?  I guess when state is restored, when reading from the earliest record, the extra data of the column name would be captured at that point?

Jackie-Jiang commented:
All the upsert metadata is re-created when the segments are loaded (when server is restarted), so it should be fine to change the config without re-ingesting the data. Note that snapshot needs to be disabled because after the config change, the invalid docs might change.

egalpin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,51 @@
+      return comparisonColumn.comparisonValue.compareTo(otherComparisonColumn.comparisonValue);
+    }
+
+    return -1;
```
Note to self: to maintain existing behaviour where the existing comparisonColumn is null and so to is the inbound comparisonColumn (which allows upsert due to compareTo returning 0 in such a case), if all inbound comparisonColumns are null, do one last check to see if all those same columns in the existing set of ComparisonColumns are also null. If so, return 0. If not, return -1

egalpin commented:
Ok @Jackie-Jiang I think I've got this now.  The change _does_ coerce `comparisonColumn` String to be a list of size one, but the semantics are such that these configurations are the same: `"comparisonColumn": "foo"` == `"comparisonColumns": ["foo"]`

In both those cases, the code paths will be "unchanged" from how they work today: no wrapping of Comparable in `ComparisonColumns`, only a single `PinotSegmentColumnReader` rather than a map when using multiple comparisonColumns, etc

egalpin commented:
Ok I'll stop changing this now haha.  Ready for another round of review.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,124 @@
+      ComparisonValue otherComparisonValue = other.getComparisonColumns().get(columnEntry.getKey());
+
+      if (otherComparisonValue == null) {
+        // This can happen if a new column is added to the list of comparisonColumns. We want to support that without
```
I don't think this is possible without server restart right now

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,124 @@
+    return -1;
+  }
+
+  private static Map<String, ComparisonValue> merge(@Nullable Map<String, ComparisonValue> current,
```
Why is `current` nullable?

Currently `merge()` happens during the `compareTo()` which is kind of anti-pattern because `compareTo()` shouldn't modify the value. Ideally we should call `merge()` separately when setting the comparison value into the `RecordLocation`.

egalpin replied:
Ya I agree, I've never been happy with the implicit merging.  I implemented this way currently so that any implementation of `BasePartitionUpsertMetadataManager` would be able to support multiple comparison columns without the need to re-implement any logic.  Ex. `RecordLocation` is an inner class of `ConcurrentMapPartitionUpsertMetadataManager`.  Each implementation of `BasePartitionUpsertMetadataManager`would need updating separately in order to ensure calling `merge` in the correct way if we make `merge` an explicit process.

Thoughts on the best way to proceed? Is `RecordLocation` or a similar abstraction something that could be made part of `BasePartitionUpsertMetadataManager` such that a factory method could be implemented and shared across extensions of the base class? Something like:

```java
  public RecordLocation updateRecordLocation(IndexSegment indexSegment, int docId, RecordInfo inbound,
      RecordLocation existing) {
    if (_comparisonColumns.size() > 1) {
      ComparisonColumns inboundComp = (ComparisonColumns) inbound.getComparisonValue();
      ComparisonColumns existingComp = (ComparisonColumns) existing.getComparisonValue();
      Map<String, Comparable> merged =
          ComparisonColumns.merge(existingComp.getComparisonColumns(), inboundComp.getComparisonColumns());
      return new RecordLocation(indexSegment, docId, new ComparisonColumns(merged));
    }

    return new RecordLocation(indexSegment, docId, inbound.getComparisonValue());
  }
```

And have sub classes all use this method?  It may not be feasible, as there might be considerable differences between sub-classes.

Jackie-Jiang reviewed (commented):
Overall logic looks good

egalpin commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertRestletResource.java:
```diff
@@ -129,14 +129,17 @@ public String estimateHeapUsage(String tableSchemaConfigStr,
+          String msg = "Not support data types for the comparison column";
+          throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+        } else {
+          bytesPerValue += dt.size();
```
@Jackie-Jiang good catch. Does this look correct now?

egalpin replied:
I suppose there's also the wrapper Object of `ComparisonColumns` which has overhead as well.

egalpin commented:
The memory overhead really starts to add up.  Now that I have the context that the upsert config cannot change without server restart, I feel that this design can be re-implemented using an array of size `_comparisonColumns.size()`, and the column names themselves would not need to be stored per PK.

@Jackie-Jiang thoughts on using an array indexed by the order of _comparisonColumns? As opposed to map implementation.

Jackie-Jiang commented:
> The memory overhead really starts to add up. Now that I have the context that the upsert config cannot change without server restart, I feel that this design can be re-implemented using an array of size `_comparisonColumns.size()`, and the column names themselves would not need to be stored per PK.
> 
> @Jackie-Jiang thoughts on using an array indexed by the order of _comparisonColumns? As opposed to map implementation.

Good point. We only need to save the values and use the column index to navigate

egalpin commented:
@Jackie-Jiang changed to array-based impl in https://github.com/apache/pinot/pull/10234/commits/e4eb4e624ebdeca30a49839d159484dd3e68cff1

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:
```diff
@@ -154,10 +135,22 @@ public void setDefaultPartialUpsertStrategy(Strategy defaultPartialUpsertStrateg
+   * in undefined behaviour. Typically, one comparisonColumn is allocated per distinct producer application of data
+   * in the case where there are multiple producers sinking to the same table.
    */
+  public void setComparisonColumns(List<String> comparisonColumns) {
```
Consider adding a check here to prevent user from setting empty list

egalpin replied:
Added, but in its current for it implies that calling this setter more than once will cause a run time error.  Is that acceptable, or better to overwrite?  If we remove these checks, then if an upsert config includes both `comparisonColumn` and `comparisonColumns`, whichever is specified last will overwrite the one specified first.

Jackie-Jiang replied:
Here I'm suggesting adding a non-empty check instead of non-null check. The `_comparisonColumns` cannot be empty
`Preconditions.checkArgument(comparisonColumns == null || !comparisonColumns.isEmpty(), ...)`

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:
```diff
@@ -558,15 +560,39 @@ private RecordInfo getRecordInfo(GenericRow row, int docId) {
   }
 
+  private RecordInfo multiComparisonRecordInfo(PrimaryKey primaryKey, int docId, GenericRow row) {
+    Comparable[] comparisonColumns = new Comparable[_upsertComparisonColumns.size()];
+
+    int i = -1;
+    for (String columnName : _upsertComparisonColumns) {
+      i++;
```
(minor) More readable
```suggestion
    int numComparisonColumns = _upsertComparisonColumns.size();
    Comparable[] comparisonValues = new Comparable[numComparisonColumns];
    for (int i = 0; i < numComparisonColumns; i++) {
      String columnName = _upsertComparisonColumns.get(i);
```

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:
```diff
@@ -558,15 +560,39 @@ private RecordInfo getRecordInfo(GenericRow row, int docId) {
+    int i = -1;
+    for (String columnName : _upsertComparisonColumns) {
+      i++;
+      Object comparisonValue = row.getValue(columnName);
+
+      Preconditions.checkState(comparisonValue instanceof Comparable,
+          "Upsert comparison column: %s must be comparable", columnName);
+
+      if (row.isNullValue(columnName)) {
+        // Inbound records may only have exactly 1 non-null value in one of the comparison column i.e. comparison
+        // columns are mutually exclusive
+        comparisonColumns[i] = null;
+        continue;
+      }
+
+      comparisonColumns[i] = (Comparable) comparisonValue;
```
Check `null` first, then read value only if it is not `null`
```suggestion
      if (!row.isNullValue(columnName)) {
        Object comparisonValue = row.getValue(columnName);
        ...
        comparisonColumns[i] = (Comparable) comparisonValue;
      }
```

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:
```diff
@@ -558,15 +560,39 @@ private RecordInfo getRecordInfo(GenericRow row, int docId) {
+          "Upsert comparison column: %s must be comparable", columnName);
+
+      if (row.isNullValue(columnName)) {
+        // Inbound records may only have exactly 1 non-null value in one of the comparison column i.e. comparison
```
Do we want to enforce this? If so, we should probably add a check

egalpin replied:
I had intended to enforce this so as to avoid highly complex comparison logic.  But we could potentially support all values being non-null if there's an agreeable way to perform the comparison.

Ex. Do all values for a new/inbound doc need to be greater than corresponding prior values? Or should it be "if any value is higher than corresponding prior value, perform the upsert?"

egalpin replied:
Might be best saved for a follow-up PR if desirable?

egalpin replied:
Added a Precondition check

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,104 @@
+
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ComparisonColumns implements Comparable<ComparisonColumns> {
+  private Comparable[] _comparisonColumns;
```
(minor) Suggest renaming it to `_values`

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/UpsertUtils.java:
```diff
@@ -75,19 +75,32 @@ public RecordInfo next() {
   public static class RecordInfoReader implements Closeable {
     public final PrimaryKeyReader _primaryKeyReader;
-    public final PinotSegmentColumnReader _comparisonColumnReader;
+    public final ComparisonColumnReader _comparisonColumnReader;
```
Let's also move `SingleComparisonColumnReader` and `MultiComparisonColumnReader` into this class as inner class

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/MultiComparisonColumnReader.java:
```diff
@@ -0,0 +1,60 @@
+  private final Map<String, PinotSegmentColumnReader> _comparisonColumnReaders;
+
+  public MultiComparisonColumnReader(IndexSegment segment, List<String> comparisonColumns) {
+    _comparisonColumnReaders = new TreeMap<>();
```
More performant if we keep a list of `PinotSegmentColumnReader`s

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,104 @@
+    int comparableIndex = getComparableIndex();
+
+    if (comparableIndex < 0) {
+      // All comparison values were null.  This record is only ok to keep if all prior values were also null
```
Is this allowed? Should we just ignore this record (return -1)

egalpin replied:
It's currently allowed in this change set in order to maintain parity with the case of a using a singular comparisonColumn.  Currently, if the comparison value is null it will take <default value> for the column.  If a subsequent record for the same PK also has a null comparison value, it will take <default value> for the column and end up with `compareTo` being `0` which results in the record being kept.

Allowing all values to be null when using multiple comparison columns would maintain this behaviour.  Thoughts?

Jackie-Jiang replied:
For singular comparison column case, comparison value should not be null, and null value can have undefined behavior because we just use the default value as the comparison value.
For multiple comparison columns case, there should be at least one comparison column not null. If one event does not have any non-null value, we can just drop it (unless there is no event associated with the key) since we don't know which value to compare.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  public static Comparable[] merge(Comparable[] current, Comparable[] next) {
+    // Create a shallow copy so {@param current} is unmodified
```
No need to create a copy here since the merge is happening in-place anyway. We can actually change the return to `void` to make it explicit that the merge is happening in-place

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,104 @@
+        comparisonResult = 1;
+      } else {
+        comparisonResult = comparisonValue.compareTo(otherComparisonValue);
+      }
```
Since we already know the `comparableIndex`, we may simplify this part to just copy other values to the current value array
```
      }
      if (comparisonResult >= 0) {
        for (int i = 0; i < _values.length; i++) {
          if (i != comparableIndex) {
            _values[i] = other._values[i];
          }
        }
      }
```

Jackie-Jiang replied:
Let's also add a TODO here to mention the modification within `compareTo()`

Jackie-Jiang reviewed (commented):
Mostly good

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,91 @@
+
+  private int getComparableIndex() {
+    for (int i = 0; i < _values.length; i++) {
+      if (_values[i] == null) {
+        continue;
+      }
+      return i;
```
(nit)
```suggestion
      if (_values[i] != null) {
        return i;
      }
```

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,91 @@
+      //  {@link BasePartitionUpsertMetadataManager}. Ideally, this merge should only be triggered explicitly by
+      //  implementations of {@link BasePartitionUpsertMetadataManager}.
+      for (int i = 0; i < _values.length; i++) {
+        // N.B. null check _must_ be here to prevent overwriting _values[i] with null from other._values[i], such
+        // as in the case where this is the first time that a non-null value has been received for a given
+        // comparableIndex after previously receiving non-null value for a different comparableIndex
```
I don't think this is possible since there can be up to one value set in the event

egalpin replied:
I added this check after my unit tests initially failed. It turned out to be a bug where the comparableIndex could change due to computing it based on first non-null value in _values and the fact that compareTo implicitly changes the contents of _values throughout execution.  Fixed in fb25fc9d38

Jackie-Jiang reviewed (approved):
LGTM with minor comments! Great job!

egalpin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:
```diff
@@ -558,15 +560,40 @@ private RecordInfo getRecordInfo(GenericRow row, int docId) {
+        comparisonValues[i] = (Comparable) comparisonValue;
+      }
+    }
+    Preconditions.checkState(numNonNull == 1 || numNonNull == 0,
```
> For multiple comparison columns case, there should be at least one comparison column not null. If one event does not have any non-null value, we can just drop it (unless there is no event associated with the key) since we don't know which value to compare.

In that case, should we make this check strictly `== 1` ?  And avoid the need to handle the case of all null Comparables[] within ComparisonColumns?

Jackie-Jiang replied:
Either way is fine. If we want to make the rule more strict, we may check if it is strict `== 1`.

egalpin replied:
Done as part of https://github.com/apache/pinot/pull/10234/commits/45066e94007f5c9aac237c14061b7107a8180092

egalpin commented:
Any outstanding items here?

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:
```diff
@@ -558,15 +560,40 @@ private RecordInfo getRecordInfo(GenericRow row, int docId) {
+      if (!row.isNullValue(columnName)) {
+        // Inbound records may only have exactly 1 non-null value in one of the comparison column i.e. comparison
+        // columns are mutually exclusive
+        numNonNull++;
```
Since we are looping over all the columns here, we can store the column index here:
```
  int comparableIndex = -1;
  for (...) {
    if (!row.isNullValue(columnName)) {
      Preconditions.checkState(comparableIndex == -1, "Must have exactly 1 non-null comparison column");
      comparableIndex = i;
    }
  }
  Preconditions.checkState(comparableIndex != -1, ...);
```

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,76 @@
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ComparisonColumns implements Comparable<ComparisonColumns> {
+  private final Comparable[] _values;
+  private int _comparableIndex;
```
We can make it final, and pass it from outside

egalpin commented on pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotUpsertRestletResource.java:
```diff
@@ -129,14 +129,25 @@ public String estimateHeapUsage(String tableSchemaConfigStr,
+          throw new ControllerApplicationException(LOGGER, msg, Response.Status.BAD_REQUEST);
+        } else {
+          if (comparisonColumns.size() == 1) {
+            bytesPerKey += dt.size();
```
Good catch, thanks!

egalpin commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:
```diff
@@ -154,10 +136,26 @@ public void setDefaultPartialUpsertStrategy(Strategy defaultPartialUpsertStrateg
+    _comparisonColumns = comparisonColumns;
+  }
+
+  public void setComparisonColumn(List<String> comparisonColumns) {
+    Preconditions.checkArgument(comparisonColumns == null || !comparisonColumns.isEmpty(),
+        "Possible duplicated entries: comparisonColumn, comparisonColumns");
+    _comparisonColumns = comparisonColumns;
+  }
+
```
Removed.  I originally had this here to allow for using either `comparisonColumn` or `comparisonColumns` interchangeably, but that's probably less clear than enforcing String with one, and List<String> for the other.

egalpin commented:
@Jackie-Jiang thanks for the re-review! I've addressed your comments in https://github.com/apache/pinot/pull/10234/commits/03423bc9ff5cc17108faed9d68ec8f17bb52e3d6

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:
```diff
@@ -154,10 +136,20 @@ public void setDefaultPartialUpsertStrategy(Strategy defaultPartialUpsertStrateg
+   * in undefined behaviour. Typically, one comparisonColumn is allocated per distinct producer application of data
+   * in the case where there are multiple producers sinking to the same table.
    */
+  public void setComparisonColumns(List<String> comparisonColumns) {
```
Just realize that certain configs might be handled unexpected. Consider the following configs:
```
{
  "comparisonColumn": "a",
  "comparisonColumns": null
}
```
```
{
  "comparisonColumn": null,
  "comparisonColumns": ["a"]
}
```
```
{
  "comparisonColumn": null
}
```
```
{
  "comparisonColumns": null
}
```

In order to handle all of them, we can modify these 2 methods in the following way:
```
  public void setComparisonColumns(List<String> comparisonColumns) {
    if (CollectionUtils.isNotEmpty(comparisonColumns) {
      _comparisonColumns = comparisonColumns;
    }
  }

  public void setComparisonColumn(String comparisonColumn) {
    if (comparisonColumn != null) {
      _comparisonColumns = Collections.singletonList(comparisonColumn);
    }
  }
```

egalpin replied:
Good catch, made those suggested changes.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java:
```diff
@@ -558,15 +560,41 @@ private RecordInfo getRecordInfo(GenericRow row, int docId) {
+            "Upsert comparison column: %s must be comparable", columnName);
+        comparisonValues[i] = (Comparable) comparisonValue;
+      }
+    }
```
We might need a precondition here to enforce one non-null value set, or it will throw exception in `ComparisonColumns.compareTo()`
```
    Preconditions.checkState(comparableIndex != -1,
          "Documents must have exactly 1 non-null comparison column value");
```

egalpin replied:
Good catch, looks like both checks are needed.  Added this check as well + more informative comment.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ComparisonColumns.java:
```diff
@@ -0,0 +1,67 @@
+  private final Comparable[] _values;
+  private final int _comparableIndex;
+
+  public ComparisonColumns(int comparableIndex, Comparable[] values) {
```
(nit)
```suggestion
  public ComparisonColumns(Comparable[] values, int comparableIndex) {
```

Jackie-Jiang commented:
Seems the github actions CI is stuck. The code looks good. Will merge once tests pass

egalpin commented:
Thanks @Jackie-Jiang !  Ya, I believe there's a github actions partial outage at the moment.

egalpin commented:
Seems GHA are back 🎉 All looks well.

Jackie-Jiang commented:
Can you help add this new feature to the documentation in this [repo](https://github.com/pinot-contrib/pinot-docs)?

egalpin commented:
sure thing @Jackie-Jiang, will do 👍
